rescript/parse_silva_taxonomy.py

```diff
@@ -147,44 +149,59 @@ def _validate_taxrank_taxmap_taxtree(prepped_taxrank, prepped_taxmap, taxtree):
                          "taxids are: ", tree_map_diffs)
 
 
-def _compile_taxonomy_output(updated_taxmap, include_species_labels=False,
-                             selected_ranks=SELECTED_RANKS):
+def _compile_taxonomy_output(updated_taxmap, ranks,
+                             include_species_labels=False):
     # Takes the updated_taxmap dataframe (i.e. the combined
     # output from _build_base_silva_taxonomy and _prep_taxmap) and only
     # returns the desired ranks (as a pandas Series) from selected_ranks, which
-    # is an ordered dict. For now, SELECTED_RANKS is used. In the future, we
-    # may provide the option to allow users to feed in their own dictionary
-    # of selected_ranks. The user can opt to return the species labels.
-    sr = list(selected_ranks.values())
+    # is an ordered dict.
+    # sort user ranks relative to allowed_ranks
+    # this ensures that the taxonomic ranks supplied by the user are in order
+    allowed_rank_prefixes = list(ALLOWED_RANKS.values())
+    user_rank_prefixes = [ALLOWED_RANKS[r] for r in ranks]
+    allowed_rank_indices = [allowed_rank_prefixes.index(r) for r in
+                            user_rank_prefixes]
+    allowed_rank_indices.sort()
+    sorted_ranks = [allowed_rank_prefixes[i] for i in allowed_rank_indices]
```

Comment:
This might be easier to understand:
```
sorted_ranks = [p for r, p in ALLOWED_RANKS.items() if r in ranks]
```

Reply:
Apparently, I still have a hang-up on list comprehensions...